tests/integration/thrift_tests.py

```diff
@@ -38,267 +42,330 @@
-            raise unittest.SkipTest("cryptography is not installed")
+        server_span_observer = mock.Mock(spec=ServerSpanObserver)
+        with serve_thrift(handler, server_span_observer) as server:
+            with raw_thrift_client(server.endpoint) as client:
+                with self.assertRaises(TestService.ExpectedException):
+                    client.example()
 
-    def test_expected_exception_not_passed_to_server_span_finish(self):
-        client_memory_trans = TMemoryBuffer()
-        client_prot = THeaderProtocolFactory().getProtocol(client_memory_trans)
-        client = TestService.Client(client_prot)
-        try:
-            client.example_throws(crash=False)
-        except (TTransportException, EOFError):
-            pass  # we don't have a test response for the client
+        server_span_observer.on_start.assert_called_once_with()
+        server_span_observer.on_finish.assert_called_once_with(None)
 
-        itrans = TMemoryBuffer(client_memory_trans.getvalue())
-        iprot = THeaderProtocolFactory().getProtocol(itrans)
-        self.processor.process(iprot, self.oprot)
+    def test_unexpected_exception_passed_to_server_span_finish(self):
+        """If the server returns an unexpected exception, mark a failure."""
 
-        self.assertEqual(self.server_observer.on_start.call_count, 1)
-        self.assertEqual(self.server_observer.on_finish.call_count, 1)
-        self.assertEqual(self.server_observer.on_finish.call_args[0], (None,))
+        class UnexpectedException(Exception):
+            pass
 
-    def test_unexpected_exception_passed_to_server_span_finish(self):
-        client_memory_trans = TMemoryBuffer()
-        client_prot = THeaderProtocolFactory().getProtocol(client_memory_trans)
-        client = TestService.Client(client_prot)
-        try:
-            client.example_throws(crash=True)
-        except (TTransportException, EOFError):
-            pass  # we don't have a test response for the client
+        class Handler(TestService.Iface):
+            def example(self, context):
+                raise UnexpectedException
+        handler = Handler()
 
-        itrans = TMemoryBuffer(client_memory_trans.getvalue())
-        iprot = THeaderProtocolFactory().getProtocol(itrans)
-        self.processor.process(iprot, self.oprot)
+        server_span_observer = mock.Mock(spec=ServerSpanObserver)
+        with serve_thrift(handler, server_span_observer) as server:
+            with raw_thrift_client(server.endpoint) as client:
+                with self.assertRaises(TApplicationException):
+                    client.example()
 
-        self.assertEqual(self.server_observer.on_start.call_count, 1)
-        self.assertEqual(self.server_observer.on_finish.call_count, 1)
-        _, captured_exc, _ = self.server_observer.on_finish.call_args[0][0]
+        server_span_observer.on_start.assert_called_once_with()
+        server_span_observer.on_finish.assert_called_once()
```

Comment:
This probably doesn't actually work, this was added to `mock` in python 3.6 https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once